File: libraries/fndf/ndf1_id2ac.f
      SUBROUTINE NDF1_ID2AC( INDF, IACB )
*+
*  Name:
*     NDF1_ID2AC

*  Purpose:
*     Convert an NDF identifier into the associated ACB index.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL NDF1_ID2AC( INDF, IACB )

*  Description:
*     The routine converts an NDF identifier, previously issued by
*     NDF1_EXPID, into an index to the appropriate entry in the ACB.
*     The identifier supplied is fully checked and a value of zero is
*     returned if it is not valid.

*  Arguments:
*     INDF = INTEGER (Given)
*        NDF identifier.
*     IACB = INTEGER (Returned)
*        Index to an entry in the ACB, or zero of the identifier
*        supplied was not valid.

*  Notes:
*     -  This routine does not perform error checking or reporting.

*  Algorithm:
*     -  Check that the INDF value supplied is positive and return a
*     value of zero if it is not.
*     -  Decode the INDF value into an index for the ACB.
*     -  Check that the identifier value matches the value originally
*     issued for the ACB slot and that the slot is still in use.
*     -  If everything is OK, then return the ACB index. Otherwise,
*     return zero.

*  Copyright:
*     Copyright (C) 1989 Science & Engineering Research Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     25-SEP-1989 (RFWS):
*        Original, derived from the equivalent ARY_ system routine.
*     {enter_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'NDF_PAR'          ! NDF_ public constants
      INCLUDE 'NDF_CONST'        ! NDF_ private constants

*  Global Variables:
      INCLUDE 'NDF_ACB'          ! NDF_ Access Control Block
*        ACB_CHK( NDF__MXACB ) = INTEGER (Read)
*           Identifier value originally issued for each ACB slot.
*        ACB_USED( NDF__MXACB ) = LOGICAL (Read)
*           Whether the ACB slot is in use.

*  Arguments Given:
      INTEGER INDF

*  Arguments Returned:
      INTEGER IACB

*  External References:
      EXTERNAL NDF1_INIT         ! Initialise common blocks

*  Local Variables:
      INTEGER I                  ! Temporary variable for ACB index

*.

*  Check that the identifier value is positive, and return a value of
*  zero if it is not.
      IF ( INDF .LE. 0 ) THEN
         IACB = 0

*  Decode the INDF value into an index for the ACB (the reverse of the
*  process used by NDF1_EXPID to encode it).
      ELSE
         I = MOD( INDF, NDF__MXACB )
         IF ( I .EQ. 0 ) I = NDF__MXACB

*  Check that the identifier matches the value originally issued for
*  this ACB slot and that the slot is still in use. If OK, then return
*  the ACB index. Otherwise, return zero.
         IF ( ( ACB_CHK( I ) .EQ. INDF ) .AND. ACB_USED( I ) ) THEN
            IACB = I
         ELSE
            IACB = 0
         END IF
      END IF

      END

File: libraries/ndf/ndf1Id2ac.c
#include <pthread.h>
#include "sae_par.h"
#include "ndf1.h"

/* Define a union which allows a bit pattern to be accessed as a
   signed or unsigned int. */
typedef union IdUnion {
   int i;
   unsigned u;
} IdUnion;

/* These globals are declared in ndf1Ffs.c */
extern int Ndf_NACB;
extern int Ndf_NPCB;
extern NdfACB **Ndf_ACB;
extern NdfPCB **Ndf_PCB;

NdfObject *ndf1Id2ac( int id, int isacb ) {
/*
*+
*  Name:
*     ndf1Id2Ac

*  Purpose:
*     Convert an identifier into the associated ACB or PCB pointer.

*  Synopsis:
*     NdfObject *ndf1Id2ac( int id, int isacb )

*  Description:
*     The routine converts an identifier (NDF or placeholder), previously
*     issued by ndf1Expid, into a pointer to the appropriate ACB pr PCB
*     structure. The identifier supplied is fully checked and a NULL pointer
*     is returned if it is not valid.

*  Parameters:
*     id
*        The integer identifier for an NDF or placeholder.
*     isacb
*        Indicates the type of object expected by the calling function.
*        Non-zero indicates an ACB is expected. Zero indicates a PCB is
*        expected.

*  Returned function value:
*     Pointer to the ACB or PCB structure, or NULL of the identifier supplied
*     was not valid.

*  Notes:
*     -  This routine does not perform error checking or reporting.

*  Copyright:
*      Copyright (C) 2018 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     DSB: David S. Berry (EAO)

*  History:
*     3-APR-2019 (DSB):
*        Original version, based on AST identifier system by RFWS.
*-
*/

/* Local variables: */
   NdfObject *result = NULL;
   IdUnion work;
   int slot;
   int nel;

/* Check an identifier was supplied. If so, wait for exclusive access to the
   ACB or PCB related global variables */
   if( isacb ) {
      if( id == NDF__NOID ) return result;
      NDF__ACB_LOCK_MUTEX;
   } else {
      if( id == NDF__NOPL ) return result;
      NDF__PCB_LOCK_MUTEX;
   }

/* Reverse the encoding process performed by ndf1Expid to retrieve the slot
   index for the ACB or PCB. */
   work.i = id;
   work.u = ( work.u ^ ( ( (unsigned) NDF__FACNO ) << 8U ) ) >> 8U;

/* Check that the offset obtained doesn't extend beyond the limits of
   the array of ACB or PCB pointers. Remember that ndf1Expid converts
   the slot number from zero-base to one-base, so we need to convert it
   back to zero-base before using it. */
   slot = work.i - 1;
   nel = isacb ? Ndf_NACB : Ndf_NPCB;
   if ( ( slot >= 0 ) && ( slot < nel ) ) {

/* Get a pointer to the ACB or PCB structure. Remember that ndf1Expid converts
   the slot number from zero-base to one-base, so we need to convert it
   back to zero-base before using it. */
      if( isacb ) {
         result = (NdfObject *) Ndf_ACB[ slot ];
      } else {
         result = (NdfObject *) Ndf_PCB[ slot ];
      }

/* See if the "check" field matches the ID value. */
      if( result->check != id ) {
         result = NULL;

/* Also check that the slot number stored in the object is the expected
   value. */
      } else if( result->slot != slot ) {
         result = NULL;

/* Also check that the slot is in used. */
      } else if( !result->used ) {
         result = NULL;
      }
   }

/* Allow other threads to access the ACB or PCB related global variables */
   if( isacb ) {
      NDF__ACB_UNLOCK_MUTEX;
   } else {
      NDF__PCB_UNLOCK_MUTEX;
   }

/* Return the result. */
   return result;
}


